contributors/design-proposals/apps/configmap-secret-gc.md

```diff
@@ -0,0 +1,271 @@
+    + [On Controller API Resource Creation](#on-controller-api-resource-creation)
+    + [On Controller API Resource Template Updates](#on-controller-api-resource-template-updates)
+- [Alternatives](#alternatives)
+  * [Embed ConfigMaps/Secrets in Controller Spec](#embed-configmapssecrets-in-controller-spec)
+  * [Specify ConfigMaps/Secrets Garbage Collection in Controller Spec](#specify-configmapssecrets-garbage-collection-in-controller-spec)
+  * [Specify ConfigMaps/Secrets garbage collection in PodSpec](#specify-configmapssecrets-garbage-collection-in-podspec)
+  * [Automatic ConfigMap/Secret References Updates](#automatic-configmapsecret-references-updates)
+- [Caveats](#caveats)
+- [Testing Plan](#testing-plan)
+
+
+## Abstract
+
+ConfigMaps allow users to decouple configuration artifacts from image content to
+keep containerized applications portable. The data stored in ConfigMaps can be
+consumed in pods or provide the configurations for system components. Secrets
+are similar to ConfigMaps, but are intended to hold sensitive information. 
+
+To do a rolling update of a ConfigMap/Secret, users need to create a new
+ConfigMap, update its references in workloads resources (such as Deployment),
+and delete the old ConfigMap. To facilitate ConfigMap/Secrets rollouts and
+management ([#22368](https://github.com/kubernetes/kubernetes/issues/22368)), we
+propose a mechanism that supports garbage collection of unused ConfigMaps and
+Secrets. 
+
+
+## Requirements
+
+1.  Garbage collection of a ConfigMap/Secret should be easy to configure,
+    manage, and understand. 
+1.  If a user wants a ConfigMap/Secret to be garbage collected, the intent
+    should be easily discoverable by other users who consume the same
+    ConfigMap/Secret. No surprises. 
+1.  The design of garbage collection of ConfigMaps/Secrets should cover
+    Deployments, StatefulSets, DaemonSets, ReplicaSets, and
+    ReplicationControllers. 
+1.  For controllers that support history and rollback feature:
+    ConfigMaps/Secrets should not be garbage collected until the referencing
+    controller history object is removed, so that rollback still works. 
+1.  Garbage collection of ConfigMaps/Secrets should work after a controller's
+    pod template is updated.
+1.  Users can only enable or disable the garbage collection of a
+    ConfigMap/Secret when creating the ConfigMap/Secret.
+
+
+## API Changes
+
+A new *immutable* API field is introduced to ConfigMap and Secret API. 
+
+
+```go
+type ConfigMap struct {
+  // If true, the supported controllers (Deployments, StatefulSets, DaemonSets,
+  // ReplicaSets, and ReplicationControllers) that reference this ConfigMap will
+  // become its owner. 
+  // Note that garbage collector will delete this ConfigMap once its owners are 
+  // gone, even if this ConfigMap is still referenced by other API resources
+  // (such as Jobs).
+  // This can only be specified at creation time.
+  // Default to false.
+  ControllersAsOwners *bool
```

Comment:
This feels yucky and introduces undesirable coupling in the system.  How will you deal with extension resources?  This won't work for extension controllers, which is a red flag.

i don't really like the idea of controllers being special on config maps.

Reply:
I agree with Clayton here. If we add field to object that means every time we'll be wanting to extend this mechanism with new resources that do not have any controller-controlled history we'll be changing APIs of those. Why not just having annotation, which would make an object eligible for GC. This will work fine with every type of object, including CRDs, etc.

Reply:
I was even thinking this could be a nice way for users to get their resources garbage collected if we combine this annotation with owerRef on each object. All we'd need to do is to enable users setting them easily. I doubt there's such a way currently. Obviously, where possible the controllers should handle those situation automatically. Then adding new resources should be pretty straightforward.

Reply:
> Why not just having annotation, which would make an object eligible for GC.

That sounds interesting. In the past, we've regretted starting things out as annotations if they eventually were destined to be fields. But maybe this is a case where it will always make sense as an annotation, like the ones that Deployment puts on ReplicaSet.

After all, this flag is not really about affecting the ConfigMap itself. It's just a signal to other entities, affecting how they treat the ConfigMap. That sounds more like an annotation than a field to me.

@janetkuo What do you think?

Reply:
@smarterclayton Extension controllers would need to implement the same logic mentioned in the "controller" section in order to support this feature. The assumption here is that users would keep creating new CMs and then update references in controller pod template. This proposal tries to help users garbage collect those kinds of CMs.

@soltysh @enisoc What does such annotation look like? If it's just a boolean value, it does not affect the design much. We need to allow users to opt in and out of CM GC, and setting a field seems more straightforward than an annotation. A field can be protected by validation as well (you cannot make an annotation immutable). 

By "combining this annotation with owerRef on each object", did you mean switching this field to an annotation, or having the garbage collector check both the annotation and ownerRef to decide if this CM should be GC'd or not? If the former, the design is the same; if the latter, we need to change the behavior of the garbage collector.

Reply:
This case is mostly needed for ConfigMap and Secret, which are "consumed" by other resources.

I don't like the name of the field, but the idea is that referrers should take ownership -- the resource should be GCed when no longer needed. It's actually a hint to consumers. It could be an annotation, but that wouldn't be very discoverable.

If we had split out the PodTemplate from controllers and used it for controller history universally, this would be easier, and could work OOTB for extensions.

As it is, this does work for extensions, but requires extension implementors to do some work.

Reply:
BTW, I believe this hint would be useful even without the history issue.

Reply:
Since it's only a hint (true means it's ok to do it, but doesn't guarantee anything), and the thing it allows is for the object to be deleted when it's no longer referenced, how about `AllowRefCount`?